### DISCUSSION TITLE
Policies & Functions Don't work

### DISCUSSION THREAD

**Discussion by Phamiliarize** [supabase/supabase#3235] [Questions] [answered]
I want to only allow creation when a user has less than a certain amount of items.

I've created a custom function here to return a boolean based on this.

```sql
create function can_make_item(auth_user uuid)
returns boolean
language plpgsql
as
$$
declare
   retval boolean;
begin
   select CAST((count(*) < 4) as boolean)
   into retval
   from items
   where items.owner_id = auth_user;
   return retval;
end;
$$;
```

When dropping this into the following policy, however, it doesn't work and seems to always be true. It looks something like this

```sql
((role() = 'authenticated'::text) AND (can_make_board(uid()) = true))
```

I don't know why but policy auto changes `auth.uid()` to just `uid()`. I thought it might be because uid does not exist and it returns nothing therefore our count would always be less than 4. When running the function in a query with a string provided ID and not the uid() function this seems to be the case, so I added a check to make sure the user is also authenticated so that if they have a UID, and it has no items, then they should be clear.

The issue is, people can still unlimitedly make objects.

Am I doing something wrong? is this functionality just not supported?

If it is supported I think these functions can be a marvelous way to helping implement some business logic.

**Comment by w3b6x9** [comment]
@Phamiliarize can you please share the entire policy you created?

**Reply by Phamiliarize** [reply]
It is an insert policy on a table called boards.
The only part I can see to edit/get from the policy is this:

(
  (role() = 'authenticated':: text)
  AND (
    can_make_board(uid()) = true
  )
)

**Comment by steve-chavez** [comment]
@Phamiliarize Your use case seems more fit for constraint triggers. You can read an article about them [here](https://www.cybertec-postgresql.com/en/triggers-to-enforce-constraints/).

Basically, you create before insert trigger(see tutorial [here](https://www.postgresqltutorial.com/creating-first-trigger-postgresql/)) and inside it you raise an exception(example [here](https://stackoverflow.com/questions/12529069/raising-error-in-postgresql)) if your constraint fails.

**Comment by Phamiliarize** [comment, marked as answer]
So I got this to work. I have two policies for creating items.

The first checks if you are in a "premium" table and does a TTL check on membership. IF that passes you can infinitely create.

The second starts with defining this function:

```

create or replace function get_free_slots(user_id uuid)
returns setof uuid as $$
   select
   owner_id
   from boards
   where owner_id = $1
   group by owner_id
   HAVING COUNT(*) < 4;

$$ stable language sql security definer;
```

Basically, you can put whatever number you want your limit to be, but make sure rather than returning true or false, you return a set with the user_id of those who are eligible. In this case our where also makes sure we only check for the user making the request. Next, the using portion of the insert policy is:

```
(uid() IN ( SELECT get_free_slots(uid()) AS get_free_slots))
```
Basically, the function won't add the user to the set if they aren't eligible, at which point our RLS will not allow the insert.
